validate.js

```diff
@@ -0,0 +1,155 @@
+    return new Error(
+      `invalid message: contentSection ${typeof contentSection} with length ${
+        contentSection.length
+      } is incorrect, expected a list of content and contentSignature`
+    )
+
+  const [content, contentSignature] = contentSection
+
+  if (
+    !(
+      content.type === 'metafeed/add/existing' ||
+      content.type === 'metafeed/add/derived' ||
+      content.type === 'metafeed/update' ||
+      content.type === 'metafeed/tombstone'
+    )
+  )
+    return new Error(
+      `invalid message: content type ${content.type} is incorrect`
+    )
+
+  const subfeedBFE = bfe.encode(content.subfeed)
+  const subfeedType = subfeedBFE.slice(0, 1).toString('hex')
+  if (subfeedType !== '00')
+    return new Error(
+      `invalid message: content subfeed type "0x${subfeedType}" is incorrect, expected 0x00`
+    )
+
+  const metafeedBFE = bfe.encode(content.metafeed)
+  const metafeedType = metafeedBFE.slice(0, 2).toString('hex')
+  if (metafeedType !== '0003')
+    return new Error(
+      `invalid message: content metafeed type "0x${metafeedType}" is incorrect, expected 0x0003`
+    )
+
+  if (content.type === 'metafeed/add/derived') {
+    if (content.nonce.length !== 32)
+      return new Error(
+        `invalid message: content nonce "${content.nonce}" is ${content.nonce.length} bytes, expected 32`
+      )
+  }
+
+  const signatureErr = validateSignature(
+    content.subfeed,
+    content,
+    contentSignature,
+    hmacKey
+  )
+  if (signatureErr) return signatureErr
+}
+
+/**
+ * Verify that the contentSignature correctly signs the content.
+ *
+ * @param {string} subfeedKey - `subfeed` key for the message
+ * @param {Buffer} content - Dictionary of meta feed metadata
+ * @param {string} contentSignature - Base64-encoded signature for the given `content`
+ * @param {Buffer | string | null} hmacKey - HMAC key that was used to sign the payload
+ * @returns {Object | undefined} Either an Error containing a message or an `undefined` value for successful verification
+ */
+function validateSignature(subfeedKey, content, contentSignature, hmacKey) {
+  const hmacKeyErr = validateHmacKey(hmacKey)
+  if (hmacKeyErr) return hmacKeyErr
+
+  const isSignatureRx = isCanonicalBase64('', '\\.sig.\\w+')
+
+  if (!isSignatureRx.test(contentSignature))
+    return new Error(
+      `invalid message: contentSignature "${contentSignature}", expected a base64 string`
+    )
+
+  const contentBFE = bfe.encode(content)
+
+  // if the subfeedKey is a supported uri, convert it to sigil for verification
+  if (!ref.isFeed(subfeedKey)) {
+    if (
+      !SSBURI.isFeedSSBURI(subfeedKey) &&
+      !SSBURI.isBendyButtV1FeedSSBURI(subfeedKey) &&
+      !SSBURI.isGabbyGroveV1FeedSSBURI(subfeedKey)
+    ) {
+      return new Error(
+        `invalid message: subfeed key "${subfeedKey}", expected a canonical uri format or classic ssb sigil`
+      )
+    } else {
+      let { type, format, data } = SSBURI.decompose(subfeedKey)
```

Comment:
```suggestion
      const { data } = SSBURI.decompose(subfeedKey)
```

Reply:
We actually don't need the sigil and the suffix, because on line 120 we're passing it to `public`, and `public` consists of only the base64 data.

Reply:
That's much neater, thanks! Wasn't quite sure how to do the selective destructuring 👍🏻 👍🏻